### ISSUE TITLE
TAP-GW: sign keys

### ISSUE THREAD

**Issue by monkeypants** [ausdigital/ausdigital-tap#12]
If the TAP-GW client is an IDP relying party (or has equivalent bearer-token due to known customer relationhip, combining contractual proof of consent with on-behalf-of relationship with the End-User Business), then the TAP-GW should be able to sign the public keys of the End-User Business to the LOA of the tokens identity claim.

How this might work:
 * client presents TAP-GW with valid bearer token
 * client POST's key signing request for public key associated with the identity claim
 * TAP-GW verifies that the public key is for an email address associated with the business identity
 * TAP-GW signs the client's public key, with a private key that corresponding to the LOA of the identity claim in the bearer token
 * TAP-GW pushes signed key to a public keyserver (P2P distribution mechanism)

This mechanism might be for a key associated with the business itself, or with keys associated with any published endpoint of the business.

There would be no need for signature revocation (it's up to the client to revoke the public key, not up to the key-signers to revoke their signatures of it).

**Comment by onthebreeze** [comment]
Need to discuss the whole question of revocation 

How will the whole framework know that Mary no longer works for acme and so her authority over that abn is gone and consequently the whole chain of trust to the vanguard token is now gone ?

Steven Capell
Mob: 0410 437854

**Comment by monkeypants** [comment]
there are two questions here.

I was asking about key revocation certificates here, for "that key is no longer trusted by me" (I lost passphrase or lost control of the private key, etc).

I think you are talking about on-behalf-of relationships; "that person, e.g. staff member or service provider, is no is longer approved to update metadata on my behalf". Maybe because you sacked them. I think at least some of those on-behalf-of relationships are described in business metadata, with LOA-escalation for resolution. In other words, I see that as a SMP/DCP requirement.

For TAP-GW, do we also need multi-user with role based authorisation as well?

**Comment by onthebreeze** [comment]
Not sure - maybe for larger orgs.  In future maybe saml auth to corporate directory 

Steven Capell
Mob: 0410 437854

**Comment by monkeypants** [comment]
But should that be in the spec?

**Comment by onthebreeze** [comment]
Good point, probs not

Steven Capell
Mob: 0410 437854

**Comment by monkeypants** [comment]
see #13 for discussion about role-based access control to TAP-GW. Bringing the focus of this ticket back to key signing, let me explain my intention here.

When a business publishes keys somewhere (could be SMP/DCP, or multiple places), it is possible for other keyholders to make an assertion "I verify the identity of this key". The collection of all these assertions form a "trust web", and there is a protocol for sharing these assertions in a peer to peer way (this is specified in RFC4880).

What I'm suggesting is that, if a TAP-GW is a client of an IDP (relying party), then the TAP-GW trusts identity claims about a business (to a certain LOA). It would be consistent with that trust level for the TAP-GW to publicly assert a verification of the key associated with the identity claim. This would cost the TAP-GW very little, and benefit the business identity by enhancing the strength of the association between their business identity and their published keys.

It would leak a little information about the customer list of the TAP-GW, so if we did support this it should probably be considered optional behaviour. The TAP-GW may or may not want to keep their customer list secret.

**Comment by monkeypants** [comment]
I think this ticket is totally out of date. We have `GET /{participant_id}/keys/` in the DCP now, returning something like this (list of signing keys for the participant identifier):
```
[
  {
    "published": "2017-03-15T07:58:00.997569+00:00",
    "keyid": "756E9B5376F8314C",
    "pubKey": "-----BEGIN PGP PUBLIC KEY BLOCK-----\nVersion: GnuPG v2\n\nmQENBFi1NKIBCACuZ6Rzqh8BLXOXEu59j5He8SGCUucDuDcna5rZ4fxi88aKDNzH\ngdEkBcd5Mneu+obe+THMzhi/9Qc6PnBdYSYyP0319vpItDLNlIPqGYJsNtKME3EO\n5WG4aie/QqCMT6IgmVZ+Im5Ph06SKkp0uJg0KPkY037z6sR3wXLl/d+dmlzfSQsC\nIykVBLMqrpbF7Cut2p9UyyK2JuiabloLNUmE9lhg62I2ZT1I+M+tCQ32wZ9zhszA\nnnGy4eCLhG7VFOPqWE8aRCPkvidetCh//XWRESSlWhBAVi75EJL8kBajz7JxKqxH\ny0Cj7QY1ddeT7lBLI7e9itueBYysc+ZZ1yXXABEBAAG0QHVybjpvYXNpczpuYW1l\nczp0YzplYmNvcmU6cGFydHlpZC10eXBlOmlzbzY1MjM6MDE1MTo6MTEwMDI4MTQ1\nNDiJATcEEwEIACEFAli1NKICGwMFCwkIBwIGFQgJCgsCBBYCAwECHgECF4AACgkQ\ndW6bU3b4MUzrfAf/WRwV+LFb+dc/cBU0XWmSwfA+/fVtEWpOntNFM5U/2R6j8vjn\nd9a5kHeJEIXJRKk5rM7PppzQ1SORZ9Jq0e/htOkRiL3bvdFeiAclX5r2NfUz49oU\n1FJj2LqrL8JzmFzeqQQb8PPgA087NmXI52XxhJQQahuJDUTof5i2dPOwOCCuAQju\nwfYjchBYyRlwICdmzdDcFJPJGG2f4BH69oI29aAwD9U7hw7kzC3YhbMOmoxCkfdP\nzJSCpYtDHeC1Fey5Yv+KMnyyap/bQYPnxFSWSg6vPe1dFGattYPx057pibI4fQvV\nZocgN2rOb8ebfq5sYUIRXAwHJWAC+1i58SbWP7kBDQRYtTSiAQgAtQjmMRy9VDQT\nFCwuPUmd94OhG/7LkvT4Q46VpbCU4g5t++Pw7YeQ38Pa4j8l/nx/eoOjdBMKE5FS\nlT2iTcoeoxSnewrWMrzyptFtwGf223rKNad/4VPRA1en9jA4qTrynMHtbse2AR1S\nVIW2tlVweooCyoKnFBubmFSowOuHoojRnehX2gvTXFW5EMb1jzHqzbMSyCusZY5t\no/coj0Bu6lTH1MkOrDwMVQ8KvR9ZRNJHzXttOct0i0zLRF87IodHfFC1+kVm6rnc\nzHgNf+GwPSDFQeU4Cy24JE9nt5ZdcKNlMiIU/AzGtyONoHR3yuF+CpN9jzk9G5ll\nbjYttzlIQwARAQABiQEfBBgBCAAJBQJYtTSiAhsMAAoJEHVum1N2+DFMHxAIAKyp\nT2XUDsjyVDRSQQSM2fqvCH71b3smTgogi48hV9ExgOO3a2QylYlmIB3xSHLtnDH6\nZBjCbVlfJiOEE8MQy4NDmrRk9ROmBRHxnHGB9KNGhOD9dL/Os69NrXnYzKqPBnQw\n86Ycj0KVH86PffBrw9D0A+eNq/58Yd0WTs/X2zGdCkB/ZwnZ8/7jGFirAWYqFVM0\np7egIw7NoRfSPr/LH3SifaDwdVwOfvRShkKV2DvAy7m/KmHFnV0faIQQsoVBBuXf\ntceoYY7cd6bEbzg0GV4zrXBQcTS0yarYpyirGq+5UCYLCyt3GCaiPG63ThRdHG4e\nsF4A27aQOiSvZUpgRYA=\n=mHW0\n-----END PGP PUBLIC KEY BLOCK-----",
    "revoked": null,
    "fingerprint": "A07CCEC0099D8C2695FB5C90756E9B5376F8314C"
  }
]
```
the combination of published date and revocation date meet our needs here.
